compiler/src/lnntoamm/Types.ts

```diff
@@ -278,84 +550,127 @@ export class FunctionType extends Type {
+          if (tys === null) {
             return fns;
           } else {
-            return [...fns, [fn, retTy] as [Fn, Type]];
+            return [...fns, [fn, ...tys] as [Fn, Type[], Type]];
           }
-        }, new Array<[Fn, Type]>()),
+        }, new Array<[Fn, Type[], Type]>()),
       );
+      isDbg && console.log('for weight now', fnsForWeight);
       fnsByWeight.set(weight, fnsForWeight);
-      if (
-        indices.every(
-          (idxInDim, dimIdx) => idxInDim === matrix[dimIdx].length - 1,
-        )
-      ) {
-        break;
-      }
-      // now change the indices. This mostly works like binary addition,
-      // except each digit `i` is in base `j` where `j = matrix[i].length`
-      // so if `matrix[1].length === 1` then the carry for `indices[1]`
-      // is just passed to `matrix[0]`
-      indices.reduceRight((carry, _matIdx, idx) => {
-        indices[idx] += carry;
-        if (indices[idx] === matrix[idx].length) {
-          indices[idx] = 0;
-          return 1;
-        } else {
-          return 0;
-        }
-      }, 1);
     }
     const weights = Array.from(fnsByWeight.keys()).sort();
     // weights is ordered lowest->highest so it's just a matter of
     // appending the tuple at each weight to a list
-    const ret = weights.reduce((fns, weight) => {
-      let weightFns = fnsByWeight.get(weight);
-      weightFns = weightFns.filter(
-        ([weightedFn, _retTy]) =>
-          fns.findIndex(([fn, _retTy]) => fn === weightedFn) === -1,
-      );
-      return [...fns, ...weightFns];
-    }, new Array<[Fn, Type]>());
-    if (ret.length > originalLength) {
+    // const ret = weights.reduce((fns, weight) => {
+    // let weightFns = fnsByWeight.get(weight);
+    // isDbg && console.log('at weight', weight, 'fns are', weightFns);
+    // weightFns.forEach(([fn, pTys, retTy]) => {
+    //   let foundIdx = fns.findIndex(([maybeSame, _p, _r]) => maybeSame === fn);
+    //   if (foundIdx !== -1) {
+    //     const [[_sameFn, samePTys, sameRetTys]] = fns.splice(foundIdx, 1);
+    //   } else {
+    //     fns.push([fn, pTys, retTy]);
+    //   }
+    // });
+    // weightFns = weightFns.filter(
+    //   ([weightedFn, _retTy]) =>
+    //     fns.findIndex(([fn, _retTy]) => fn === weightedFn) === -1,
+    // );
+    // isDbg && console.log('filtered:', weightFns);
+    // return [...fns, ...weightFns];
+    const ret: [Fn[], Type[][], Type[]] = weights.reduce(
+      ([fns, pTys, retTys], weight) => {
+        fnsByWeight
+          .get(weight)
+          .forEach(([weightedFn, weightedPTys, weightedRetTy]) => {
+            const alreadyIdx = fns.findIndex((fn) => fn === weightedFn);
+            if (alreadyIdx === -1) {
+              // push to the end of the fns since technically the weight for the function
+              // is indeed higher than what it was before
+              fns.push(weightedFn);
+            } else {
+              fns.push(fns.splice(alreadyIdx, 1)[0]);
```

Comment:
What the heck is this doing?

Reply:
it's moving the function from the position with the lower weight to the position with the higher weight. the comment for the other branch should be for the `else` branch